src/common/io.py

```diff
@@ -1593,23 +1610,29 @@ def get_variables_data(self,
         if isinstance(start_index, int) and isinstance(stop_index, int) and stop_index < start_index:
             raise ValueError(f"Invalid values for {start_index=} and {stop_index=}, " + \
                               "'start_index' needs to be less than or equal to 'stop_index'.")
-        trajectories = []
 
-        # Get the time trajectory
+        trajectories = {}
+
+        # Get the corresponding time trajectory
         if not self._contains_diagnostic_data:
             time = self._get_trajectory(0, start_index, stop_index)
         else:
             # Since we interpolate data if diagnostics is enabled
             time = self._get_diagnostics_trajectory(0, start_index, stop_index)
 
-        # Need to account for data that might be added while we are iterating over 'names' later
-        if stop_index is None:
-            stop_index = len(time) + start_index
+        # If stop_index > number of data points, and data gets added while we are iterating
+        # then we might get trajectories of unequal lengths. Therefore ensure we set stop_index here accordingly.
+        stop_index = min(len(time) + start_index, float('inf') if stop_index is None else stop_index)
 
         for name in names:
-            trajectories.append(self._get_variable_data_as_trajectory(name, time, start_index, stop_index))
+            trajectories[name] = self._get_variable_data_as_trajectory(name, time, start_index, stop_index)
 
-        new_start_index = start_index + len(time) if len(trajectories) > 0 else None
+        largest_trajectory_length = -1
+        for v, t in trajectories.items():
+            largest_trajectory_length = max(largest_trajectory_length, len(t.x))
+        new_start_index = (start_index + largest_trajectory_length) if trajectories else start_index
```

Comment:
@chria Note the changes on line 1633

Reply:
Edge case question: What happens if you invoke the function for start_index = 0, stop_index = None, for a parameter and a variable, when there is only the initial data point available?

Is this a rare case where the new start index would be incorrectly adjusted for the parameter trajectory length?

Reply:
In this case there is no result file yet. I just tested this with `fmu.initialize()`, but the result file has not been generated yet so therefore I don't think we have to worry about this edge case. Let me know if you disagree!

Reply:
I wouldn't say this is nothing to worry about, if you start simulation (after initialization) + polling for new solution points at the same time, you will very likely run into this. You'd then miss one trajectory point.

When exactly is the result file created then?

Reply:
It looks like it is done exactly when the simulation starts.

I did another test `res = fmu.simulate(0, 0)`. Here we have `len(res['v'])` equal to 1 (bouncing ball example). Below is the output with different input:
```
# Input: start_index = 1, stop_index = 5
next_start_index=1
Len for variable='h': len(data.x)=0
Len for variable='g': len(data.x)=0
Len for variable='v': len(data.x)=0
# Input: start_index = 451, stop_index = 651
next_start_index=451
Len for variable='h': len(data.x)=0
Len for variable='g': len(data.x)=0
Len for variable='v': len(data.x)=0
# Input: start_index = 0, stop_index = None
next_start_index=1
Len for variable='h': len(data.x)=1
Len for variable='g': len(data.x)=1
Len for variable='v': len(data.x)=1
```
I would say the results are as expected, right?

Reply:
I agree that this is something that we should fix. The "next index" should only consider continuous variables. As for testing, do a test instead using the specific calls to the ResultHandler routines.

Reply:
> It looks like it is done exactly when the simulation starts.
> 
> I did another test `res = fmu.simulate(0, 0)`. Here we have `len(res['v'])` equal to 1 (bouncing ball example). Below is the output with different input:
> 
> ```
> # Input: start_index = 1, stop_index = 5
> next_start_index=1
> Len for variable='h': len(data.x)=0
> Len for variable='g': len(data.x)=0
> Len for variable='v': len(data.x)=0
> # Input: start_index = 451, stop_index = 651
> next_start_index=451
> Len for variable='h': len(data.x)=0
> Len for variable='g': len(data.x)=0
> Len for variable='v': len(data.x)=0
> # Input: start_index = 0, stop_index = None
> next_start_index=1
> Len for variable='h': len(data.x)=1
> Len for variable='g': len(data.x)=1
> Len for variable='v': len(data.x)=1
> ```
> 
> I would say the results are as expected, right?

This depends on the exact call and variables.

If there is only a single data point, then `get_variables_data([<parameter>, <continuous variable>], 0, None)` is the interesting case.

Reply:
Please see 51228a7

Reply:
A minor minor comments there, but looks good overall!